aqt/metadata.py

```diff
@@ -0,0 +1,657 @@
+        modules_ver: Optional[str] = None,
+        extensions_ver: Optional[str] = None,
+        architectures_ver: Optional[str] = None,
+        tool_name: Optional[str] = None,
+        tool_long_listing: Optional[str] = None,
+    ):
+        """
+        Construct ListCommand.
+
+        :param filter_minor:        When set, the ListCommand will filter out all versions of
+                                    Qt that don't match this minor version.
+        :param is_latest_version:   When True, the ListCommand will find all versions of Qt
+                                    matching filters, and only print the most recent version
+        :param modules_ver:         Version of Qt for which to list modules
+        :param extensions_ver:      Version of Qt for which to list extensions
+        :param architectures_ver:   Version of Qt for which to list architectures
+        """
+        self.logger = getLogger("aqt.archives")
+        self.archive_id = archive_id
+        self.filter_minor = filter_minor
+
+        if archive_id.is_tools():
+            if tool_name:
+                self.request_type = "tool variant names"
+                self._action = lambda: self.fetch_tool_modules(tool_name)
+            elif tool_long_listing:
+                self.request_type = "tool long listing"
+                self._action = lambda: self.fetch_tool_long_listing(tool_long_listing)
+            else:
+                self.request_type = "tools"
+                self._action = self.fetch_tools
+        elif is_latest_version:
+            self.request_type = "latest version"
+            self._action = lambda: Versions(self.fetch_latest_version())
+        elif modules_ver:
+            self.request_type = "modules"
+            self._action = lambda: self.fetch_modules(self._to_version(modules_ver))
+        elif extensions_ver:
+            self.request_type = "extensions"
+            self._action = lambda: self.fetch_extensions(
+                self._to_version(extensions_ver)
+            )
+        elif architectures_ver:
+            self.request_type = "architectures"
+            self._action = lambda: self.fetch_arches(
+                self._to_version(architectures_ver)
+            )
+        else:
+            self.request_type = "versions"
+            self._action = self.fetch_versions
+
+    def action(self) -> Union[List[str], Versions]:
+        return self._action()
+
+    def run(self) -> int:
+        try:
+            output = self.action()
+            if not output:
+                self.logger.info(
+                    "No {} available for this request.".format(self.request_type)
+                )
+                self.print_suggested_follow_up(self.logger.info)
+                return 1
+            if self.archive_id.is_tools():
+                print("\n".join(output))
+            elif isinstance(output, Versions):
+                print(f"{output}")
+            else:
+                print(" ".join(output))
```

Comment:
I would be very careful about this kind of change. The Azure Pipeline is failing in this section (https://dev.azure.com/miurahr/github/_build/results?buildId=4344&view=logs&j=3b077b59-19b3-5c3c-d9ea-4f2d359714ca&t=fcf4a360-d878-547e-2092-53786c55870d&l=56), because output is of type `ListCommand.Table`, and `"\n".join(output)` doesn't work on that object type.

If you're going to use this approach to printing output, you will need to account for that case as well.

Reply:
It use format when object supports, otherwise it use `" ".join(output)` which uses iterator of output.
It is better that classes supports iterator but Table does not support.